synapse/storage/user_directory.py

```diff
@@ -30,7 +28,162 @@
+
+        # Get all the rooms that we want to process.
+        def _make_staging_area(txn):
+            sql = (
+                "CREATE TABLE IF NOT EXISTS "
+                + TEMP_TABLE
+                + "(room_id TEXT NOT NULL, events BIGINT NOT NULL)"
+            )
+            txn.execute(sql)
+
+            # Get rooms we want to process from the database
+            sql = """
+                SELECT room_id, count(*) FROM current_state_events
+                GROUP BY room_id
+            """
+            txn.execute(sql)
+            rooms = [{"room_id": x[0], "events": x[1]} for x in txn.fetchall()]
+
+            self._simple_insert_many_txn(txn, TEMP_TABLE, rooms)
+
+            progress["stage"] = "processing"
+            self._background_update_progress_txn(
+                txn, "populate_user_directory", progress
+            )
+
+        def _delete_staging_area(txn):
+            sql = "DROP TABLE IF EXISTS " + TEMP_TABLE
+            txn.execute(sql)
+
+        if progress.get("stage") != "processing":
+            new_pos = yield self.get_max_stream_id_in_current_state_deltas()
+            progress["pos"] = new_pos
+            yield self.runInteraction(
+                "populate_user_directory_temp_build", _make_staging_area
+            )
+
+        def _get_next_batch(txn):
+            sql = """
+                SELECT room_id FROM %s
+                ORDER BY events DESC
+                LIMIT %s
+            """ % (
+                TEMP_TABLE,
+                str(batch_size),
+            )
+            txn.execute(sql)
+            rooms_to_work_on = txn.fetchall()
+
+            if not rooms_to_work_on:
+                return None
+
+            rooms_to_work_on = [x[0] for x in rooms_to_work_on]
+
+            sql = "SELECT COUNT(*) FROM " + TEMP_TABLE
+            txn.execute(sql)
+            progress["remaining"] = txn.fetchone()[0]
+
+            return rooms_to_work_on
+
+        rooms_to_work_on = yield self.runInteraction(
+            "populate_user_directory_temp_read", _get_next_batch
+        )
+
+        if not rooms_to_work_on:
+            yield self.runInteraction(
+                "populate_user_directory_temp_cleanup", _delete_staging_area
+            )
+            yield self.update_user_directory_stream_pos(progress["pos"])
+            yield self._end_background_update("populate_user_directory")
+            defer.returnValue(1)
+
+        logger.info(
+            "Processing the next %d rooms of %d remaining"
+            % (len(rooms_to_work_on), progress["remaining"])
+        )
+
+        for room_id in rooms_to_work_on:
+            is_in_room = yield self.is_host_joined(room_id, self.server_name)
+
+            if is_in_room:
+                is_public = yield self.is_room_world_readable_or_publicly_joinable(
+                    room_id
+                )
+
+                users_with_profile = yield state.get_current_user_in_room(room_id)
```

Comment:
Can we use `self.get_users_in_room` please, as that uses the `current_state_events` table, rather than using the state handler to recalculate it.

Reply:
as discussed: this doesn't include the profile. sob